libdevice/device_imf.hpp

```diff
@@ -0,0 +1,428 @@
+  }
+  // corner case: subnormal -> normal
+  // The denormal number of FP16 can be represented by FP32, therefore we need
+  // to recover the exponent and recalculate the fration.
+  if (__builtin_expect(exp16 == 0 && frac16 != 0, 0)) {
+    uint8_t offset = 0;
+    do {
+      ++offset;
+      frac16 <<= 1;
+    } while ((frac16 & 0x400) != 0x400);
+    // mask the 9th bit
+    frac16 &= 0x3ff;
+    exp32 = 113 - offset;
+  }
+
+  uint32_t frac32 = frac16 << 13;
+
+  uint32_t fp32_bits = 0;
+  fp32_bits |= sign;
+  fp32_bits |= (exp32 << 23);
+  fp32_bits |= frac32;
+  return *reinterpret_cast<float *>(&fp32_bits);
+#elif defined(__SPIR__)
+  return __spirv_FConvert_Rfloat_rte(x);
+#endif
+}
+
+class _iml_half {
+public:
+  _iml_half(_iml_half_internal h) : _half_internal(h) {}
+  _iml_half() = default;
+  _iml_half(const _iml_half &) = default;
+  _iml_half &operator=(const _iml_half &rh) = default;
+  _iml_half &operator=(float fval) {
+    _half_internal = __float2half(fval);
+    return *this;
+  }
+  _iml_half(float fval) : _half_internal(__float2half(fval)) {}
+  explicit operator float() const { return __half2float(_half_internal); }
+
+  _iml_half_internal get_internal() const { return _half_internal; }
+  bool operator==(const _iml_half &rh) {
+    return _half_internal == rh._half_internal;
+  }
+  bool operator!=(const _iml_half &rh) { return !operator==(rh); }
+#if defined(__SYCL_DEVICE_ONLY__) && (__SPIR__)
+  _iml_half &operator+=(const _iml_half &rh) {
+    _half_internal += rh._half_internal;
+    return *this;
+  }
+  _iml_half &operator-=(const _iml_half &rh) {
+    _half_internal -= rh._half_internal;
+    return *this;
+  }
+  _iml_half &operator*=(const _iml_half &rh) {
+    _half_internal *= rh._half_internal;
+    return *this;
+  }
+  _iml_half &operator/=(const _iml_half &rh) {
+    _half_internal /= rh._half_internal;
+    return *this;
+  }
+  _iml_half &operator++() {
+    _half_internal += 1;
+    return *this;
+  }
+  _iml_half operator++(int) {
+    _iml_half res(*this);
+    operator++();
+    return res;
+  }
+  _iml_half &operator--() {
+    _half_internal -= 1;
+    return *this;
+  }
+  _iml_half operator--(int) {
+    _iml_half res(*this);
+    operator--();
+    return res;
+  }
+
+  _iml_half operator-() {
+    _iml_half res(-_half_internal);
+    return res;
+  }
+
+  bool operator<(const _iml_half &rh) {
+    return _half_internal < rh._half_internal;
+  }
+  bool operator>(const _iml_half &rh) {
+    return _half_internal > rh._half_internal;
+  }
+#else
+  _iml_half &operator+=(const _iml_half &rh) {
+    *this = (operator float() + static_cast<float>(rh));
+    return *this;
+  }
+  _iml_half &operator-=(const _iml_half &rh) {
+    *this = (operator float() - static_cast<float>(rh));
+    return *this;
+  }
+  _iml_half &operator*=(const _iml_half &rh) {
+    *this = (operator float() * static_cast<float>(rh));
+    return *this;
+  }
+  _iml_half &operator/=(const _iml_half &rh) {
```

Comment:
Have you considered support of hosts with native float16 hardware instructions (SapphireRapids)? Maybe it should be one more branch here.